File: app/src/main/java/com/billpayments/features/payments/ui/ReceiptScreen.kt
package com.billpayments.features.payments.ui

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Button
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBar
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.billpayments.R
import com.billpayments.core.format.formatDate
import com.billpayments.core.format.formatPaise
import java.time.ZoneId

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun ReceiptScreen(viewModel: ReceiptViewModel, language: String, onDone: () -> Unit) {
    val data by viewModel.screenData.collectAsStateWithLifecycle()
    Scaffold(topBar = { TopAppBar(title = { Text(stringResource(R.string.payment)) }) }) { padding ->
        Box(modifier = Modifier.fillMaxSize().padding(padding), contentAlignment = Alignment.Center) {
            when (val state = data) {
                ReceiptScreenData.NotFound -> Text(stringResource(R.string.payment_not_found))
                is ReceiptScreenData.Processing -> ProcessingContent(state, language)
                is ReceiptScreenData.Success -> SuccessContent(state, language, viewModel::saveBiller, onDone)
                is ReceiptScreenData.Failed -> FailedContent(state, language, viewModel::retryPayment, onDone)
            }
        }
    }
}

@Composable
private fun ProcessingContent(data: ReceiptScreenData.Processing, language: String) {
    Column(horizontalAlignment = Alignment.CenterHorizontally, verticalArrangement = Arrangement.spacedBy(16.dp)) {
        CircularProgressIndicator()
        Text(stringResource(R.string.paying_amount_to, formatPaise(data.amountPaise, language), data.billerName))
    }
}

@Composable
private fun SuccessContent(
    data: ReceiptScreenData.Success,
    language: String,
    onSaveBiller: (String) -> Unit,
    onDone: () -> Unit,
) {
    Column(
        modifier = Modifier.fillMaxSize().padding(24.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(12.dp, Alignment.CenterVertically),
    ) {
        Text(formatPaise(data.amountPaise, language), style = MaterialTheme.typography.headlineLarge)
        Text(stringResource(R.string.paid_to, data.billerName))
        Text(stringResource(R.string.receipt_number, data.paymentId))
        Text(stringResource(R.string.account_value, data.account))
        Text(
            stringResource(
                R.string.date_value,
                formatDate(data.paidAt.atZone(ZoneId.systemDefault()).toLocalDate(), language),
            ),
        )
        if (data.canSaveBiller) {
            OutlinedButton(onClick = { onSaveBiller(data.billerName) }, modifier = Modifier.fillMaxWidth()) {
                Text(stringResource(R.string.save_biller))
            }
        }
        Button(onClick = onDone, modifier = Modifier.fillMaxWidth()) { Text(stringResource(R.string.done)) }
    }
}

@Composable
private fun FailedContent(
    data: ReceiptScreenData.Failed,
    language: String,
    onRetry: () -> Unit,
    onDone: () -> Unit,
) {
    Column(
        modifier = Modifier.fillMaxSize().padding(24.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(12.dp, Alignment.CenterVertically),
    ) {
        Text(
            stringResource(R.string.payment_failed_summary, formatPaise(data.amountPaise, language), data.billerName),
            style = MaterialTheme.typography.titleMedium,
        )
        Button(onClick = onRetry, modifier = Modifier.fillMaxWidth()) { Text(stringResource(R.string.retry_payment)) }
        OutlinedButton(onClick = onDone, modifier = Modifier.fillMaxWidth()) { Text(stringResource(R.string.back_to_home)) }
    }
}

File: lib/features/payments/ui/receipt_screen.dart
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:go_router/go_router.dart';

import '../../../core/format/formats.dart';
import '../../../l10n/l10n.dart';
import '../../saved_billers/data/saved_biller.dart';
import '../../saved_billers/data/saved_billers_provider.dart';
import '../data/payments_provider.dart';
import 'receipt_screen_data.dart';
import 'receipt_screen_data_provider.dart';

class ReceiptScreen extends ConsumerWidget {
  const ReceiptScreen({super.key, required this.paymentId});

  final String paymentId;

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final data = ref.watch(receiptScreenDataProvider(paymentId));
    return Scaffold(
      appBar: AppBar(title: Text(ref.l10n.payment)),
      body: Padding(
        padding: const EdgeInsets.all(16),
        child: switch (data) {
          ReceiptLoading() =>
            const Center(child: CircularProgressIndicator()),
          ReceiptNotFound() => Center(child: Text(ref.l10n.paymentNotFound)),
          ReceiptProcessing() => _Processing(data: data),
          ReceiptSuccess() => _Success(data: data),
          ReceiptFailed() => _Failed(data: data),
        },
      ),
    );
  }
}

class _Processing extends ConsumerWidget {
  const _Processing({required this.data});

  final ReceiptProcessing data;

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    return Center(
      child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          const CircularProgressIndicator(),
          const SizedBox(height: 16),
          Text(ref.l10n.payingAmountTo(
              formatPaise(data.amountPaise, ref.languageCode),
              data.billerName)),
        ],
      ),
    );
  }
}

class _Success extends ConsumerWidget {
  const _Success({required this.data});

  final ReceiptSuccess data;

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      children: [
        const SizedBox(height: 24),
        Icon(Icons.check_circle,
            size: 64, color: Theme.of(context).colorScheme.primary),
        const SizedBox(height: 16),
        Text(formatPaise(data.amountPaise, ref.languageCode),
            textAlign: TextAlign.center,
            style: Theme.of(context).textTheme.headlineMedium),
        Text(ref.l10n.paidTo(data.billerName), textAlign: TextAlign.center),
        const SizedBox(height: 16),
        Card(
          child: Padding(
            padding: const EdgeInsets.all(16),
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Text(ref.l10n.receiptNumber(data.paymentId)),
                Text(ref.l10n.accountValue(data.account)),
                Text(ref.l10n.dateValue(
                    formatDate(data.paidAt, ref.languageCode))),
              ],
            ),
          ),
        ),
        const Spacer(),
        if (data.canSaveBiller)
          OutlinedButton.icon(
            onPressed: () => ref.read(savedBillersProvider.notifier).save(
                SavedBiller(
                    billerId: data.billerId,
                    account: data.account,
                    nickname: data.billerName)),
            icon: const Icon(Icons.bookmark_add),
            label: Text(ref.l10n.saveBiller),
          ),
        const SizedBox(height: 8),
        FilledButton(
          onPressed: () => context.go('/'),
          child: Text(ref.l10n.done),
        ),
      ],
    );
  }
}

class _Failed extends ConsumerWidget {
  const _Failed({required this.data});

  final ReceiptFailed data;

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      children: [
        const SizedBox(height: 24),
        Icon(Icons.error, size: 64, color: Theme.of(context).colorScheme.error),
        const SizedBox(height: 16),
        Text(
            ref.l10n.paymentFailedSummary(
                formatPaise(data.amountPaise, ref.languageCode),
                data.billerName),
            textAlign: TextAlign.center,
            style: Theme.of(context).textTheme.titleMedium),
        const Spacer(),
        FilledButton(
          onPressed: () {
            final paymentId = ref.read(paymentsProvider.notifier).pay(
                billerId: data.billerId,
                billerName: data.billerName,
                categoryId: data.categoryId,
                account: data.account,
                amountPaise: data.amountPaise);
            if (paymentId != null) context.go('/payment/$paymentId');
          },
          child: Text(ref.l10n.retryPayment),
        ),
        const SizedBox(height: 8),
        TextButton(
          onPressed: () => context.go('/'),
          child: Text(ref.l10n.backToHome),
        ),
      ],
    );
  }
}
